sdk/vaa/structs.go

```diff
@@ -367,11 +408,161 @@ func Unmarshal(data []byte) (*VAA, error) {
 	return v, nil
 }
 
+func UnmarshalBatch(data []byte) (*BatchVAA, error) {
+	if len(data) < minBatchVAALength {
+		return nil, fmt.Errorf("BatchVAA.Observation is too short")
+	}
+	v := &BatchVAA{}
+
+	v.Version = data[0]
+	if v.Version != BatchVAAVersion {
+		return nil, fmt.Errorf("unsupported VAA version: %d", v.Version)
+	}
+
+	reader := bytes.NewReader(data[1:])
+
+	if err := binary.Read(reader, binary.BigEndian, &v.GuardianSetIndex); err != nil {
+		return nil, fmt.Errorf("failed to read guardian set index: %w", err)
+	}
+
+	lenSignatures, er := reader.ReadByte()
+	if er != nil {
+		return nil, fmt.Errorf("failed to read signature length")
+	}
+
+	v.Signatures = make([]*Signature, int(lenSignatures))
+	for i := 0; i < int(lenSignatures); i++ {
+		index, err := reader.ReadByte()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read validator index [%d]", i)
+		}
+
+		signature := [65]byte{}
+		if n, err := reader.Read(signature[:]); err != nil || n != 65 {
+			return nil, fmt.Errorf("failed to read signature [%d]: %w", i, err)
+		}
+
+		v.Signatures[i] = &Signature{
+			Index:     uint8(index),
+			Signature: signature,
+		}
+	}
+
+	lenHashes, err := reader.ReadByte()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read hashes length [%w]", err)
+	}
+	v.Hashes = make([]common.Hash, int(lenHashes))
+	for i := 0; i < int(lenHashes); i++ {
+		hash := [32]byte{}
+		if n, err := reader.Read(hash[:]); err != nil || n != 32 {
+			return nil, fmt.Errorf("failed to read hash [%d]: %w", i, err)
+		}
+		v.Hashes[i] = common.BytesToHash(hash[:])
+	}
+
+	lenObservations, err := reader.ReadByte()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read observations length: %w", err)
+	}
+
```

Comment:
The length of hashes and observations should always be in sync. Even if they were different here, the guardian does not do anything with hashes or observations after it has signed and broadcast the BatchVAA to gossip.

I added two data integrity checks to this function anyway, just because it doesn't hurt, and it's nice to have canaries.